src/lib/tree/data-source/flat-data-source.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {CollectionViewer, DataSource} from '@angular/cdk/collections';
+import {FlatTreeControl, TreeControl} from '@angular/cdk/tree';
+import {Observable} from 'rxjs/Observable';
+import {merge} from 'rxjs/observable/merge';
+import {map} from 'rxjs/operators/map';
+import {take} from 'rxjs/operators/take';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
+
+/**
+ * Tree flattener to convert a normal type of node to node with children & level information.
+ *
+ * For example, the input data is nested:
+ *   SomeNode: {
+ *     key: 'Fruits',
+ *     children: [
+ *       NodeOne: {
+ *         key: 'Apple',
+ *       },
+ *       NodeTwo: {
+ *        key: 'Pear',
+ *      }
+ *    ]
+ *  }
+ *  After flattener flatten the tree, the structure will become
+ *  SomeNode: {
+ *    key: 'Fruits',
+ *    expandable: true,
+ *    level: 1
+ *  },
+ *  NodeOne: {
+ *    key: 'Apple',
+ *    expandable: false,
+ *    level: 2
+ *  },
+ *  NodeTwo: {
+  *   key: 'Pear',
+  *   expandable: false,
+  *   level: 2
+  * }
+ */
+export class MatTreeFlattener<T, F> {
```

Comment:
I don't quite follow what `T` and `F` are. Is one the type of node and one the type of data? If so, should the type of node include an `extends`?

Reply:
Discussed offline and added comments about this.